### PR TITLE
bug: ensure settings from file are filtered like env settings

### PR DESCRIPTION
**Pull request by jrconlin**
There is a bug where ADM settings read from a file would return early and NOT receive the additional checks. 

This extends https://github.com/mozilla-services/contile/pull/311

### CODE REVIEW

File: src/adm/settings.rs [hunk 1]
```diff
@@ -168,26 +177,46 @@ impl From<&mut Settings> for AdmSettings {
...
+                if let Some(ref paths) = filter.paths {
+                    return paths
+                        .iter()
+                        .any(|path| path.matching == "prefix" && !path.value.ends_with('/'));
+                }
+                false
+            }) {
+                return Err(ConfigError::Message(format!("Advertiser {:?} advertiser_urls contain invalid prefix PathFilter (missing trailing '/')", adv)));
             }
             if filter_setting.advertiser_urls.iter().any(|filter| {
```
Comment (ncloudioj) [inline src/adm/settings.rs:221]: Did you want to remove this block now as it already returns an error above?

File: src/adm/settings.rs [hunk 2]
```diff
@@ -168,26 +177,46 @@ impl From<&mut Settings> for AdmSettings {
...
+                return Err(ConfigError::Message(format!(
+                    "Advertiser {:?} include_regions must be uppercase",
+                    adv
+                )));
+            }
+            if filter_setting.advertiser_urls.iter().any(|filter| {
+                if let Some(ref paths) = filter.paths {
+                    return paths
+                        .iter()
+                        .any(|path| path.matching == "prefix" && !path.value.ends_with('/'));
```
Comment (ncloudioj) [inline src/adm/settings.rs:215]: While you're refactoring the code here, could you also add a check for "exact" values? Specifically, valid "exact" values should start with '/'.

Comment (ncloudioj) [reply src/adm/settings.rs:215]: To make things easier, I think we can just use a single error message to capture both, such as `Advertiser {:?} advertiser_urls contain invalid path values`

**Review by ncloudioj — APPROVED** [review]
Looks great, r+ with review comments.

**Review by pjenvey — CHANGES_REQUESTED** [review]

File: src/adm/settings.rs [hunk 3]
```diff
@@ -168,26 +177,46 @@ impl From<&mut Settings> for AdmSettings {
...
                 .iter()
                 .any(|region| region != &region.to_uppercase())
             {
-                panic!("Advertiser {:?} include_regions must be uppercase", adv);
+                return Err(ConfigError::Message(format!(
+                    "Advertiser {:?} include_regions must be uppercase",
+                    adv
+                )));
+            }
+            if filter_setting.advertiser_urls.iter().any(|filter| {
```
[the remark this answers is not in the document]
Comment (jrconlin) [reply src/adm/settings.rs:243]: removed duplication.

File: src/adm/settings.rs [hunk 4]
```diff
@@ -59,12 +63,44 @@ pub struct AdvertiserUrlFilter {
     pub(crate) paths: Option<Vec<PathFilter>>,
 }
 
+#[derive(Clone, Debug, Deserialize, Serialize)]
```
Comment (pjenvey) [inline src/adm/settings.rs:66]: nit: could add `Copy` as well

File: src/adm/settings.rs [hunk 5]
```diff
@@ -59,12 +63,44 @@ pub struct AdvertiserUrlFilter {
...
+            "exact" => Ok(Self::Exact),
+            _ => Err(ConfigError::Message(format!(
+                "Invalid Path Filter Type {}",
+                string
+            ))),
+        }
+    }
+}
+
+impl From<PathMatching> for String {
```
Comment (pjenvey) [inline src/adm/settings.rs:88]: total nit because I don't even think this is used but this impl could be `for &'static str` to avoid the allocation.

**Review by pjenvey — DISMISSED** [review]
r+ w/ a couple nits on the new enum

File: src/adm/settings.rs (hunk 5 above)
Comment (jrconlin) [reply src/adm/settings.rs:88]: Yeah, I do these out of habit because I inevitably need them later.
Thanks for the tip, though. I'll do that.

**Review by pjenvey — APPROVED** [review]

**Pull request merged**